main.rb

```diff
@@ -0,0 +1,111 @@
+# frozen_string_literal: true
+
+module Enumerable
+  def my_each
+    return to_enum unless block_given?
+
+    arr = to_a
+
+    i = 0
+    while i < size
+      yield(arr[i])
+      i += 1
+    end
+  end
+
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    arr = to_a
+
+    i = 0
+    while i < size
+      yield(arr[i], i)
+      i += 1
+    end
+  end
+
+  def my_select
+    return to_enum unless block_given?
+
+    result = []
+    my_each do |elem|
+      result << elem if yield elem
+    end
+    result
+  end
+  # rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+
+  def my_all?(arg = nil)
+    if block_given?
+      my_each { |elem| return false unless yield(elem) }
+    elsif arg.class == Class
+      my_each { |elem| return false unless elem.class.ancestors.include? arg }
+    elsif arg.class == Regexp
+      my_each { |elem| return false unless elem =~ arg }
+    elsif arg.nil?
+      my_each { |elem| return false unless elem }
+    else
+      my_each { |elem| return false unless elem == arg }
+    end
+  end
```

Comment:
- [ ]  Enumerables `my_all?` returns true if the block never returns false or nil
`[1,2,3,4,5].my_all?{ |num| num < 10 }` gives `nil` instead of `true`

- [ ]  Enumerables `my_all?` when no block or argument is given returns true when none of the collection members are false or nil
`[1,2,3,4,5].my_all?` should return `true`

- [ ] Enumerables `my_all?` when a class is passed as an argument returns true if all of the collection is a member of such class
`[1,2,3,4,5].my_all?(Integer)` should return `true`

- [ ] Enumerables `my_all?` when a Regex is passed as an argument returns true if all of the collection matches the Regex
`['saheed', 'oladele', 'suretrust'].my_all?(/d/)` should return `true`

- [ ] Enumerables `my_all?` when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
`[1,2,3,4,5].my_all?(3)` should return `false`